File: chicago-crime/src/main/scala/ChicagoCrime/SparkSupport.scala
package ChicagoCrime

import org.apache.spark.SparkContext
import org.apache.spark.serializer.KryoSerializer
import org.apache.spark.sql.{SQLContext, SparkSession}
import org.datasyslab.geospark.serde.GeoSparkKryoRegistrator
import org.datasyslab.geosparksql.utils.GeoSparkSQLRegistrator
import org.datasyslab.geosparkviz.sql.utils.GeoSparkVizRegistrator

trait SparkSupport {
  val spark: SparkSession = SparkSession.builder
    .appName("chicago-crime")
    .master(sys.env("SPARK_MASTER"))
    .config("spark.serializer", classOf[KryoSerializer].getName)
    .config("spark.kryo.registrator", classOf[GeoSparkKryoRegistrator].getName)
    .getOrCreate()

  @transient val sc: SparkContext = spark.sparkContext
  val sqlContext: SQLContext      = spark.sqlContext

  private def init(): Unit = {
    GeoSparkSQLRegistrator.registerAll(spark)
    GeoSparkVizRegistrator.registerAll(spark)

    sc.setLogLevel("INFO")
    sc.hadoopConfiguration.set("fs.s3a.access.key", sys.env("AWS_ACCESS_KEY_ID"))
    sc.hadoopConfiguration.set("fs.s3a.secret.key", sys.env("AWS_SECRET_ACCESS_KEY"))
    sc.hadoopConfiguration.set("fs.s3a.fast.upload", "true")
  }

  init()

  def close(): Unit = {
    spark.close()
  }
}

File: iot-device/src/main/python/spark_support.py
import logging
import os

from pyspark import SparkContext
from pyspark.sql.session import SparkSession


class SparkSupport:

    def __init__(self):
        spark = SparkSession.builder \
            .master(os.getenv("SPARK_MASTER")) \
            .getOrCreate()
        self.spark = spark
        self.sc = spark.sparkContext
        logging.info("Spark driver version: " + self.sc.version)

    def get_spark_context(self) -> SparkContext:
        return self.sc

    def get_spark_session(self) -> SparkSession:
        return self.spark
